### ISSUE TITLE
[BUG] Return value (nested model) not filtered according to response_model 

### ISSUE THREAD

**Issue by chbndrhnns** [tiangolo/fastapi#800]
### Describe the bug

I have a `Response` wrapper for my responses that only has a `data` field which contains the actual response data.
This actual data is not filtered according to the specified `response_model` unless I explicitly convert it into the desired response_model.

### To Reproduce

Steps to reproduce the behavior with a minimum self-contained file.

Replace each part with your own scenario:

1. Create a file with:

```Python
from typing import TypeVar, Generic, Optional, List

from fastapi import FastAPI
from pydantic import BaseModel
from pydantic.generics import GenericModel
from requests import Response
from starlette import status
from starlette.testclient import TestClient

DataT = TypeVar('DataT')


class R(GenericModel, Generic[DataT]):
    """Wrapper for responses"""
    data: Optional[DataT]


class ProjectOut(BaseModel):
    id: int


class ProjectInDB(ProjectOut):
    user_id: int


app = FastAPI()


@app.get(
    '/simple',
    response_model=ProjectOut
)
def get_simple() -> ProjectOut:
    project = ProjectInDB(id=1, user_id=2)

    return project


@app.get(
    '/nested_explicit',
    response_model=R[ProjectOut]
)
def get_nested() -> R[ProjectOut]:
    project = ProjectInDB(id=1, user_id=2)

    return R[ProjectOut](data=ProjectOut(**project.dict()))


@app.get(
    '/nested',
    response_model=R[ProjectOut]
)
def get_nested() -> R[ProjectOut]:
    project = ProjectInDB(id=1, user_id=2)

    return R[ProjectOut](data=project)


def test_response_model():
    with TestClient(app) as client:
        response: Response = client.get('/simple')

    assert response.status_code == status.HTTP_200_OK

    data: dict = response.json()
    assert 'user_id' not in data.keys()


def test_nested_response_model_explicit():
    with TestClient(app) as client:
        response: Response = client.get('/nested_explicit')

    assert response.status_code == status.HTTP_200_OK

    response_wrapper: dict = response.json()
    data = response_wrapper['data']
    assert 'user_id' not in data.keys()


def test_nested_response_model():
    with TestClient(app) as client:
        response: Response = client.get('/nested')

    assert response.status_code == status.HTTP_200_OK

    response_wrapper: dict = response.json()
    data = response_wrapper['data']
    assert 'user_id' not in data.keys()
```

3. Run the tests with pytest: the test `test_nested_response_model` fails

### Expected behavior

When returning data, it is filtered according to the specified `response_model`.

### Environment

- OS: macOS
- FastAPI Version: 0.45.0
- Python version: 3.7.5

### Additional context

Might this be related to https://github.com/tiangolo/fastapi/issues/675?

**Comment by chbndrhnns** [comment]
I also tried changing my wrapper class to this and it still fails:

```python
class R(GenericModel, Generic[DataT]):
    """Wrapper for responses"""
    data: Optional[ProjectOut]
```

**Comment by dmontagu** [comment]
Well this is an interesting bug; it actually passes for me on older versions of FastAPI/pydantic so I think it might be an issue with the 1.0 upgrade.

Thank you for putting together an easily-reproduced test case, I'll see if I can figure it out.

**Comment by dmontagu** [comment]
I found the issue: it's because `dict(model)` now no longer does a deep conversion of a model into its dict, only the top layer, and this is used by `BaseModel.validate`.

I see various possible fixes:
1) Change `BaseModel.validate` in pydantic to use `value_as_dict = value.dict()` instead of `value_as_dict = dict(value)`. I think this is non-preferable since the current approach is faster and may be preferable in a non-FastAPI context.
2) Change FastAPI to perform the dict conversion itself. There are various potential ways one could do this, but most would probably involve substantial overhead for dumping and reparsing (maybe that's similar to what's happening now anyway though). For example, adding `response = jsonable_encoder(response)` would solve this completely, though would likely be somewhat slower than the (old, working version of the) current approach.
3) Fix the secure cloned field creation: (it seems like it's not recursing down to submodels and cloning them as well; that didn't matter before because pydantic recursively converted the whole object to dict-ified form).
4) Implement `dump_as_type` in pydantic, and use it to get rid of the reparsing and secured cloned field entirely. This is my strong preference, since it will have very large performance benefits beyond fix this bug, and would provide a means for other frameworks to much more simply leverage similar funcitonality, but it presents a substantial addition to pydantic and may take some time/effort. I opened a PR for this a while ago but it was a little too large in scope; I've added some of the other pieces to pydantic since then so now could be a good time to add the rest.


Also, in case there is any confusion, this has nothing to do with generics -- it fails in the same way if you replace `R[ProjectOut]` with `RProjectOut` for
```python
class RProjectOut(BaseModel):
    """Wrapper for responses"""
    data: Optional[ProjectOut]
```
